### PR TITLE
[2/n] Modify registerCamper endpoints with Stripe creation

### PR DESCRIPTION
**Pull request by e-wai**
## Notion ticket link

[Ticket Name](https://www.notion.so/uwblueprintexecs/Task-Board-db95cd7b93f245f78ee85e3a8a6a316d)

## Implementation description
* 

## Steps to test
1.

## What should reviewers focus on?
* 

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/typescript/utilities/stripeUtils.ts [hunk 1]
```diff
@@ -90,6 +92,83 @@ export async function createStripePrice(
...
+): Stripe.Checkout.SessionCreateParams.LineItem[] => {
+  return sessionsToRegister.flatMap((campSession) => {
+    const priceItems = [
+      { price: campSession.campPriceId, quantity: campers.length },
+    ];
+
+    // every camper belongs to every session
+    // all same EDLP price
+    // campers have different EDLP dates, which may belong to different sessions
+    // we want to isolate EDLP per session
```
Comment (e-wai) [inline backend/typescript/utilities/stripeUtils.ts:109]: todo remove this

File: backend/typescript/utilities/stripeUtils.ts [hunk 2]
```diff
@@ -90,6 +92,83 @@ export async function createStripePrice(
...
+          const lpDates = campSession.dates.filter((date) =>
+            camperLPDates.has(date.toString()),
+          );
+          camperLateTotal += lpDates.length;
+        }
+
+        return [earlyTotal + camperEarlyTotal, lateTotal + camperLateTotal];
+      },
+      [0, 0],
+    );
```
Comment (e-wai) [inline backend/typescript/utilities/stripeUtils.ts:136]: this is wrong -- identifies # of days, not # of EDLP slots. correct in later PR

File: backend/typescript/services/implementations/camperService.ts [hunk 3]
```diff
@@ -97,10 +103,12 @@ class CamperService implements ICamperService {
         {},
         { session },
       );
+
       // Ensure all session ids passed in are found in the db
       if (sessionsToRegister.length !== campSessions.length) {
         throw new Error("Not all session ids passed in were found");
       }
+
```
Comment (dhmi19) [inline backend/typescript/services/implementations/camperService.ts:111]: I think I forgot to check if the camp is a published camp which is not completed. Could you add the camp status check (there should be a get status function somewhere in utils- also idk why a camp doesn't just have a status but here we are)

File: backend/typescript/services/implementations/camperService.ts [hunk 4]
```diff
@@ -144,10 +183,10 @@ class CamperService implements ICamperService {
             latePickup: camper.latePickup,
             specialNeeds: camper.specialNeeds,
             contacts: camper.contacts,
-            registrationDate: camper.registrationDate,
-            hasPaid: camper.hasPaid,
+            registrationDate: new Date().toString(),
+            hasPaid: false,
             formResponses: camper.formResponses,
-            chargeId: camper.chargeId,
+            chargeId: createStripeCheckoutSessionResponse.id,
```
Comment (dhmi19) [inline backend/typescript/services/implementations/camperService.ts:189]: ahh gotcha, ignore my comment in the last PR

File: backend/typescript/types.ts [hunk 5]
```diff
@@ -85,7 +90,7 @@ export type CamperCSVInfoDTO = {
   "Requires Early Drop-off": string;
   "Requires Late Pick-up": string;
   // eslint-disable-next-line
-  "Allergies": string;
+  Allergies: string;
```
Comment (dhmi19) [inline backend/typescript/types.ts:93]: Why are we making this change?

File: backend/typescript/utilities/stripeUtils.ts (hunk 1 above)
Comment (dhmi19) [reply backend/typescript/utilities/stripeUtils.ts:109]: If you keep this logic, I think it's worth having explanation comments here

File: backend/typescript/utilities/stripeUtils.ts (hunk 2 above)
Comment (dhmi19) [reply backend/typescript/utilities/stripeUtils.ts:136]: If we only care about the number of EDLP slots, then we can always just compare against the camp start and end time right? All sessions in the camp will start and end at the same time. So we might not care about which session the EDLP slot belongs to if we can just use the start and end time to calculate the number of slots for ED and LP. 

Also the camper model has 2 separate array fields for ED and LP so you could use those to identify if you need start or end time

File: backend/typescript/services/implementations/camperService.ts [hunk 6]
```diff
@@ -128,8 +136,39 @@ class CamperService implements ICamperService {
...
+        !campers.every(
+          (camper) =>
+            camper.age >= camp.ageLower && camper.age <= camp.ageUpper,
+        )
+      ) {
+        throw new Error("Not all campers meet the age requirements");
+      }
+
+      // Add the total charges for the campers:
+      const lineItems = createStripeLineItems(
```
Comment (dhmi19) [inline backend/typescript/services/implementations/camperService.ts:150]: do we need these line items later at some point or can we access them through the checkout session?

**Review by dhmi19 — APPROVED** [review]
Couple of comments but the logic seems ok to me

File: backend/typescript/utilities/stripeUtils.ts (hunk 2 above)
Comment (dhmi19) [reply backend/typescript/utilities/stripeUtils.ts:136]: nvm I think you did that in the next PR

Comment (e-wai) [reply backend/typescript/utilities/stripeUtils.ts:136]: yeah I need another PR to fix a lot of the EDLP logic

File: backend/typescript/services/implementations/camperService.ts (hunk 6 above)
Comment (e-wai) [reply backend/typescript/services/implementations/camperService.ts:150]: Not sure we'll need them elsewhere?

File: backend/typescript/utilities/stripeUtils.ts (hunk 1 above)
Comment (e-wai) [reply backend/typescript/utilities/stripeUtils.ts:109]: yeah this will disappear soon

File: backend/typescript/types.ts (hunk 5 above)
Comment (e-wai) [reply backend/typescript/types.ts:93]: This was an automatic linter change for me

**Pull request merged**